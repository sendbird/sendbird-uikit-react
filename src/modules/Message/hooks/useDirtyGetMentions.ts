import { useEffect, useLayoutEffect, useState } from 'react';

import { Logger } from '../../../lib/SendbirdState';
import { getMentionNodes } from '../utils/getMentionNodes';

interface DynamicParams {
  ref: React.RefObject<HTMLElement>;
}

interface StaticParams {
  logger: Logger;
}

/**
 * This is a dirty way to get the mentions given DOM node
 */
export function useDirtyGetMentions({
  ref,
}: DynamicParams, {
  logger,
}: StaticParams) {
  // Select the node that will be observed for mutations
  const targetNode = ref.current as HTMLElement;
  const [mentionNodes, setMentionNodes] = useState<Element[]>([]);

  // to get the initial mentions
  useLayoutEffect(() => {
    if (targetNode) {
      const mentions = getMentionNodes(targetNode);
      setMentionNodes(mentions);
    }
  }, [targetNode]);

  useEffect(() => {
    // Options for the observer (which mutations to observe)
    const config = { childList: true, subtree: true };

    // Callback function to execute when mutations are observed
    const callback = (mutationList: MutationRecord[]): void => {
      const hasMutation = mutationList.length > 0;
      if (hasMutation) {
        setMentionNodes(getMentionNodes(targetNode));
      }
    };

    // Create an observer instance linked to the callback function
    const observer = new MutationObserver(callback);

    if (targetNode) {
      // Start observing the target node for configured mutations
      observer.observe(targetNode, config);
      logger.info('useDirtyGetMentions: observer started', { observer, config });
    }

    return () => {
      try {
        observer.disconnect();
        logger.info('useDirtyGetMentions: observer disconnected', { observer });
      } catch (error) {
        logger.error('useDirtyGetMentions: observer disconnect failed', { observer });
      }
    };
  }, [targetNode]);
  return mentionNodes;
}
