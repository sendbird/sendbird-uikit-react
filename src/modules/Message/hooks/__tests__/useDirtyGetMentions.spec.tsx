import { render } from '@testing-library/react';
import React, { useRef } from 'react';
import { useDirtyGetMentions } from '../useDirtyGetMentions';

const logger = {
  info: jest.fn(),
  warning: jest.fn(),
  error: jest.fn(),
};

const TestComponent = ({
  mentions,
  showInput,
}: {
  mentions: number[],
  showInput: boolean,
}): JSX.Element => {
  const ref = useRef<HTMLDivElement>(null);
  const mentionNodes = useDirtyGetMentions({ ref }, { logger });
  return (
    <div>
      <div id="mentions">
        {mentionNodes.map((node) => (
          <div key={node.id}>{node.id}</div>
        ))}
      </div>
      {
        showInput && (
          <div ref={ref} id="input">
            {
              mentions.map((mention) => (
                <div data-sb-mention='true' id={mention.toString()} key={mention.toString()}>
                  mention-{mention}
                </div>
              ))
            }
          </div>
        )
      }
    </div>
  );
};

describe('useDirtyGetMentions', () => {
  it('should get mention nodes when Input is visible', () => {
    const { asFragment } = render(<TestComponent mentions={[1, 2, 3]} showInput={true} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should get empty list of mentions when Input is not rendered', () => {
    const { asFragment } = render(<TestComponent mentions={[1, 2, 3]} showInput={false} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should get list of mentions when Input is changed', () => {
    const { rerender, asFragment } = render(<TestComponent mentions={[1, 2, 3]} showInput={false} />);
    expect(asFragment()).toMatchSnapshot();
    rerender(<TestComponent mentions={[1, 2, 3, 4]} showInput={true} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
