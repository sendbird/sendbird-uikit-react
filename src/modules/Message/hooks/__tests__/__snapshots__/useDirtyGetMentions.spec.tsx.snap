// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDirtyGetMentions should get empty list of mentions when Input is not rendered 1`] = `
<DocumentFragment>
  <div>
    <div
      id="mentions"
    />
  </div>
</DocumentFragment>
`;

exports[`useDirtyGetMentions should get list of mentions when Input is changed 1`] = `
<DocumentFragment>
  <div>
    <div
      id="mentions"
    />
  </div>
</DocumentFragment>
`;

exports[`useDirtyGetMentions should get list of mentions when Input is changed 2`] = `
<DocumentFragment>
  <div>
    <div
      id="mentions"
    />
    <div
      id="input"
    >
      <div
        data-sb-mention="true"
        id="1"
      >
        mention-1
      </div>
      <div
        data-sb-mention="true"
        id="2"
      >
        mention-2
      </div>
      <div
        data-sb-mention="true"
        id="3"
      >
        mention-3
      </div>
      <div
        data-sb-mention="true"
        id="4"
      >
        mention-4
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`useDirtyGetMentions should get mention nodes when Input is visible 1`] = `
<DocumentFragment>
  <div>
    <div
      id="mentions"
    />
    <div
      id="input"
    >
      <div
        data-sb-mention="true"
        id="1"
      >
        mention-1
      </div>
      <div
        data-sb-mention="true"
        id="2"
      >
        mention-2
      </div>
      <div
        data-sb-mention="true"
        id="3"
      >
        mention-3
      </div>
    </div>
  </div>
</DocumentFragment>
`;
